### PR TITLE
Add support for pre-created VPC in VPCNetworkConfiguration CR

### PR DESCRIPTION
**Pull request by wenyingd**
1. Skip NSX VPC creations if VPC field is set in the VPCNetworkConfiguration CR
    spec.
2. Using the pre-created VPC's private IPs. connectivity profile and other
    necessary settings to set NetworkInfo CR status.
3. Add tag "nsx/managed-by: nsx-op" on the NSX VPCs which are created by
    nsx-operator.

### CODE REVIEW

File: pkg/nsx/services/vpc/vpc.go [hunk 1]
```diff
@@ -781,3 +808,44 @@ func (vpcService *VPCService) getLBProvider() string {
 	}
 	return LBProviderAVI
 }
+
+// ListNamespacesWithPreCreatedVPC returns a mapping of the pre-created VPC path and the Namespaces
+// which use the VPC. The key is the pre-created VPC path, the value is a slice of Namespaces that
+// use the VPC.
+func (service *VPCService) ListNamespacesWithPreCreatedVPC() map[string][]string {
```
Comment (zhengxiexie) [inline pkg/nsx/services/vpc/vpc.go:815]: where is it used?

Comment (wenyingd) [reply pkg/nsx/services/vpc/vpc.go:815]: My fault, originally I have added a separate goroutine to periodically read the pre-created VPC's settings and update to NetworkInfo CR, and I removed that part from this change. It seems I didn't remove code clean. Will remove it.

File: pkg/nsx/services/vpc/vpc.go [hunk 2]
```diff
@@ -781,3 +808,44 @@ func (vpcService *VPCService) getLBProvider() string {
...
+	vpcNSMappings := make(map[string][]string)
```
[the remark this answers is not in the document]
Comment (wenyingd) [reply pkg/nsx/services/vpc/vpc.go:816]: See https://github.com/vmware-tanzu/nsx-operator/pull/683#discussion_r1721243382

File: pkg/nsx/services/vpc/vpc.go [hunk 3]
```diff
@@ -143,6 +143,10 @@ func (s *VPCService) GetVPCNetworkConfigByNamespace(ns string) *common.VPCNetwor
 // TBD: for now, if network config info do not contains private cidr, we consider this is
 // incorrect configuration, and skip creating this VPC CR
 func (s *VPCService) ValidateNetworkConfig(nc common.VPCNetworkConfigInfo) bool {
+	if nc.VPCPath != "" {
+		// if network config is using a pre-created VPC, skip the check on PrivateIPs.
```
[the remark this answers is not in the document]
Comment (wenyingd) [reply pkg/nsx/services/vpc/vpc.go:147]: updated.

File: pkg/nsx/services/vpc/vpc.go [hunk 4]
```diff
@@ -781,3 +808,44 @@ func (vpcService *VPCService) getLBProvider() string {
...
+			vpcNSMappings[vpcPath] = existingNamespaces
+		}
+	}
+	return vpcNSMappings
+}
+
+func (service *VPCService) GetVPCFromNSXByPath(vpcPath string) (*model.Vpc, error) {
+	vpcResInfo, err := common.ParseVPCResourcePath(vpcPath)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (wenyingd) [reply pkg/nsx/services/vpc/vpc.go:816]: added.

**Review by zhengxiexie — APPROVED** [review]

**Review by TaoZou1 — DISMISSED** [review]

File: pkg/controllers/networkinfo/networkinfo_utils.go [hunk 5]
```diff
@@ -42,21 +43,31 @@ func deleteSuccess(r *NetworkInfoReconciler, _ *context.Context, o *v1alpha1.Net
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, common.MetricResTypeNetworkInfo)
 }
 
-func setNetworkInfoVPCStatus(ctx *context.Context, networkInfo *v1alpha1.NetworkInfo, client client.Client, createdVPC *v1alpha1.VPCState) {
+func setNetworkInfoVPCStatus(ctx *context.Context, networkInfo *v1alpha1.NetworkInfo, client client.Client, createdVPC *v1alpha1.VPCState) bool {
```
Comment (heypnus) [inline pkg/controllers/networkinfo/networkinfo_utils.go:46]: Why it returns a bool value, it seems that nobody consumes the returned value?

File: pkg/nsx/services/vpc/vpc.go [hunk 6]
```diff
@@ -781,3 +808,23 @@ func (vpcService *VPCService) getLBProvider() string {
 	}
 	return LBProviderAVI
 }
+
+func (service *VPCService) GetVPCFromNSXByPath(vpcPath string) (*model.Vpc, error) {
+	vpcResInfo, err := common.ParseVPCResourcePath(vpcPath)
+	if err != nil {
+		log.Error(err, "failed to parse VPCResourceInfo from the given VPC path %s", vpcPath)
```
Comment (heypnus) [inline pkg/nsx/services/vpc/vpc.go:815]: It seems that the logger doesn't support the way likes printf.

File: pkg/controllers/networkinfo/networkinfo_utils.go (hunk 5 above)
Comment (wenyingd) [reply pkg/controllers/networkinfo/networkinfo_utils.go:46]: I planned to use the "bool" value to mark if the CR is actually updated or not, and then the outer-caller can use this result to decide if  the metric needs a update accordingly. Unfortunately, I removed my consumer out of this change, and I didn't update the existing callers (`deleteFail, updateFail, updateSuccess`). Maybe I can revert the change?

File: pkg/nsx/services/vpc/vpc.go (hunk 6 above)
Comment (wenyingd) [reply pkg/nsx/services/vpc/vpc.go:815]: updated.

**Review by dantingl — COMMENTED** [review]
commet message: pre-craeted -> pre-created

File: pkg/controllers/namespace/namespace_controller.go [hunk 7]
```diff
@@ -210,7 +210,6 @@ func (r *NamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			log.Error(err, "failed to build namespace and network config bindings", "Namepspace", ns)
 			return common.ResultRequeueAfter10sec, nil
 		}
-		// read annotation "nsx.vmware.com/shared_vpc_namespace", if ns contains this annotation, it means it will share infra VPC
```
[the remark this answers is not in the document]
Comment (wenyingd) [reply pkg/controllers/namespace/namespace_controller.go:213]: The comment is incorrect, having confirmed with @seanpang-vmware , it is not required any more after his refactor.

**Review by heypnus — DISMISSED** [review]

**Review by dantingl — APPROVED** [review]

**Review by dantingl — APPROVED** [review]

**Review by heypnus — APPROVED** [review]

**Pull request merged**